ros/angel_system_nodes/angel_system_nodes/berkeley_task_monitor.py

```diff
@@ -0,0 +1,355 @@
+            Image,
+            self._topic_output_images_topic,
+            10  # TODO: Learn QoS meanings
+        )
+        self._task_graph_service = self.create_service(
+            QueryTaskGraph,
+            self._query_task_graph_topic,
+            self.query_task_graph_callback
+        )
+
+        # Get the task info from the config
+        with open(self._config_file, 'r') as f:
+            config = yaml.safe_load(f)
+
+        self._task_title = config["title"]
+        log.info(f"Task: {self._task_title}")
+
+        self._steps = [i["description"] for i in config["steps"]]
+
+        self._completed_steps = np.array(np.zeros(len(self._steps)), dtype=bool)
+
+        # Step classifier
+        berkeley_root = "angel_system/berkeley"
+        parser = model.get_parser()
+
+        config_file = f"{berkeley_root}/configs/MC50-InstanceSegmentation/mask_rcnn_R_101_FPN_1x_demo.yaml"
+        conf_thr = 0.7
+        args = parser.parse_args(f"--config-file {config_file} --confidence-threshold {conf_thr}".split())
+
+        log.info("Arguments: " + str(args))
+
+        cfg = model.setup_cfg(args)
+
+        os.environ['CUDA_VISIBLE_DEVICES'] = '0'
+        self.demo = predictor.VisualizationDemo_add_smoothing(
+            cfg,
+            last_time=2,
+            draw_output=self._draw_output
+        )
+
+        self.idx = 0
+
+        # Represents the current state of the task
+        self._current_step = None
+        self._current_step_id = 0
+        self._previous_step = None
+        self._previous_step_id = 0
+
+        # Control thread access to advancing the task step
+        self._task_lock = threading.RLock()
+
+        # Start the keyboard monitoring thread
+        log.info(f"Starting keyboard threads")
+        self._keyboard_t = threading.Thread(target=self.monitor_keypress)
+        self._keyboard_t.daemon = True
+        self._keyboard_t.start()
+        log.info(f"Starting keyboard threads... done")
+
+        # Publish task update to indicate the initial state
+        self.publish_task_state_message()
+
+    def query_task_graph_callback(self, request, response):
+        """
+        Populate the `QueryTaskGraph` response with the task list
+        and return it.
+        """
+        log = self.get_logger()
+        task_g = TaskGraph()
+        
+        # TODO: support different task levels?
+        task_g.task_steps = self._steps
+        task_g.task_levels = [0] * len(self._steps)
+
+        response.task_graph = task_g
+
+        response.task_title = self._task_title
+        return response
+
+    def listener_callback(self, image):
+        """
+        """
+        #import time
+        #time.sleep(2)
+        log = self.get_logger()
+        self.idx += 1
+
+        # Convert ROS img msg to CV2 image
+        bgr_image = BRIDGE.imgmsg_to_cv2(image, desired_encoding="bgr8")
+
+        predictions, step_infos, visualized_output = self.demo.run_on_image_smoothing_v2(
```

Comment:
Should we be passing `draw_output=False` when `self._draw_output == False`?

Reply:
sorry, where? self._draw_output is passed to `demo` when it is initialized

Reply:
I see it now, I don't remember how I missed that. Resolved!